hoomd/data/data_structures.py

```diff
@@ -0,0 +1,497 @@
+        typing = _get_inner_typeconverter(type_def, TypeConverterMapping)
+        return HOOMDDict(typing, parent, data, label)
+    elif isinstance(data, MutableSequence):
+        typing = _get_inner_typeconverter(type_def, TypeConverterSequence)
+        return HOOMDList(typing, parent, data, label)
+    elif isinstance(data, MutableSet):
+        return HOOMDSet(typing, parent, data, label)
+    else:
+        return data
+
+
+class HOOMDList(MutableSequence, _HOOMDDataStructures):
+    """List with type validation.
+
+    Use `to_base` to get a plain `list`.
+
+    Warning:
+        Should not be instantiated by users.
+    """
+    def __init__(self, type_definition, parent, initial_value=None,
+                 label=None):
+        self._type_definition = type_definition
+        self._parent = parent
+        self._list = []
+        self._label = label
+        self._buffered = False
+        if initial_value is not None:
+            for type_def, val in zip(cycle(type_definition), initial_value):
+                self._list.append(_to_hoomd_data_structure(val, type_def, self))
+
+    def __getitem__(self, index):
+        """Get the value(s) associated with entry index of the list.
+
+        Supports slicing.
+        """
+        return self._list[index]
+
+    def __setitem__(self, index, value):
+        """Set an item(s) at index to the given value.
+
+        Support assignment with slices.
+        """
+        if isinstance(index, slice):
+            with self._buffer():
+                for i, v in zip(self._slice_to_iterable(index), value):
+                    self[i] = v
+        else:
+            type_def = self._get_type_def(index)
+            try:
+                validated_value = type_def(value)
+            except TypeConversionError as err:
+                raise TypeConversionError(
+                    "Error in setting item {} in list.".format(index)) from err
+            else:
+                # If we override a list item that is itself a
+                # _HOOMDDataStructures we disconnect the old structure to
+                # prevent updates from it to force sync the list.
+                if isinstance(self._list[i], _HOOMDDataStructures):
+                    self._list[i]._parent = None
+
+                self._list[i] = _to_hoomd_data_structure(
+                    validated_value, type_def, self)
+        self._update()
+
+    def _slice_to_iterable(self, slice_):
+        start, end, step = slice_.start, slice_.end, slice_.step
+        if start is None:
+            if step is not None and step < 0:
+                iterable_start = len(self) - 1
+            else:
+                iterable_start = 0
+        if end is None:
+            if step is not None and step < 0:
+                iterable_end = -1
+            else:
+                iterable_end = len(self)
+        step = 1 if step is None else step
+        return range(iterable_start, iterable_end, step)
```

Comment:
I think this whole function is equivalent to:
```suggestion
    def _slice_to_iterable(self, slice_):
        return range(len(self))[slice_]
```